Flutter SDK updates (#1097)
rudderlabs/rudderstack-docs: docs/sources/event-streams/sdks/rudderstack-flutter-sdk/flutter-v2.mdx

  color={'blue'}
  style={'flat'}
/>

<GhBadge
  url={'https://pub.dev/packages/rudder_sdk_flutter'}
  label={'stability'}
  message={'beta'}
  color={'blue'}
/>

Refer to the Flutter SDK's <a href="https://github.com/rudderlabs/rudder-sdk-flutter">GitHub Repository</a> for the implementation-specific details.

## SDK setup requirements

1. It is recommended to set up the [Flutter development environment](https://flutter.dev/docs/get-started/install) in your system.
2. Sign up to [RudderStack Cloud](https://app.rudderstack.com/signup).
3. <Link to="/dashboard-guides/sources/#adding-a-source">Set up a Flutter source</Link> in the dashboard. You should be able to see the <strong>write key</strong> for this source, as shown:

<img src="../../../../assets/event-stream-sources/flutter-write-key-new.png" alt="Flutter source write key" />

You will also need the data plane URL. Refer to the <Link to="/dashboard-guides/overview/#data-plane-url">Dashboard Overview</Link> guide for more information on the data plane URL and where to find it.

## Installing the RudderStack Flutter SDK

<div class="infoBlock">

> Starting from v1.0.2, the RudderStack Flutter SDK is migrated to Null Safety.

Reviewer: Lil confused.. Google says Null Safety is a feature but here it sounds like a package manager like maven central, nuget or pub.

Author: It's this feature: https://www.section.io/engineering-education/null-safety-support-for-flutter-and-dart/#:~:text=Null%20safety%20means%20that%20a,are%20shown%20at%20compile%2Dtime. cc @desusai7 for confirmation.

Author: maybe we can reword it to something like:

```
Starting from v1.0.2, the RudderStack Flutter SDK supports the <a href="https://dart.dev/null-safety">Null Safety</a> feature.
```

Reviewer 2: yes, it's like a standard or a feature, which we were not adhering to till version 1.0.2.

And also the term migrating to null safety makes sense because we need to make many changes in the code base to be null safety compatible. 

Refer to the Flutter official doc for Migrating to Null Safety here https://dart.dev/null-safety/migration-guide

Author: Let's use the original "...migrated to Null Safety." only, then

To install and use the Flutter SDK in your web app, follow the above steps to add the Flutter SDK. Additionally, you need to <Link to="/sources/event-streams/sdks/rudderstack-javascript-sdk/quick-start-guide/#step-1-install-javascript-sdk">install the RudderStack script</Link> in the `<head>` section of your web page.

> Not all the Flutter SDK API are applicable for the web. Calling the following API on your web app will have **no effect**:

Reviewer: ```suggestion
Not all the Flutter SDK APIs are applicable for the web. For example, the following APIs will have **no effect** on your web app :
```

| `writeKey` | String       | Required      | Your Flutter source write key.                       |
| `config`   | `RudderConfig` | Optional       | Contains the RudderStack client configuration. |

## Self-hosted control plane

> For the <Link to="/destinations/rudderstack-connection-modes/#device-mode">device mode</Link>-supported destinations, the Flutter SDK needs to fetch the required configuration from the control plane. 

Reviewer: Does this mean that this section is only relevant for device mode-supported destinations?

Author: Nope, it's for the open source deployments where even the control plane is hosted. In that case, the SDK needs to know where the source-destination config file is hosted.

Reviewer: understood.. then why do we say "for device mode-supported destinations"?

Author: Makes sense - removed the sentence entirely to avoid any confusion.

> If you are self-hosting RudderStack and using the <a href="https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/">Control Plane Lite</a> utility to host your own control plane, then follow the steps in [this section](https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/#using-sdk-sources-set-up-in-self-hosted-control-plane) and specify `controlPlaneUrl` in your [`RudderConfig.Builder`](#sdk-initialization-options) that points to the hosted source configuration file.

Reviewer: ```suggestion
If you are self-hosting RudderStack using the <a href="https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/">Control Plane Lite</a> utility, then follow the steps in [this section](https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/#using-sdk-sources-set-up-in-self-hosted-control-plane) and specify the `controlPlaneUrl` in your [`RudderConfigBuilder`](#sdk-initialization-options) that points to the hosted source configuration file.
```

print(context["traits"]);
```

> For the web apps, the `getRudderContext` API returns the user's `traits` and the `anonymousId` only.

Reviewer: ```suggestion
For the web apps, the `getRudderContext` API returns the user's `traits` and the `anonymousId`.
```

## Setting your own `anonymousId` using `putAnonymousId`

> By default, RudderStack uses the `deviceId` as the user's `anonymousId`. Use the following method to override and set your own `anonymousId` with the SDK:

Reviewer: ```suggestion
By default, RudderStack uses the `deviceId` as the user's `anonymousId`. Use the following method to override and set your own `anonymousId`:
```

```

<div class="infoBlock">

> The <code class="inline-code">id</code> parameter that you pass in the <code class="inline-code">putAdvertisingId</code> method is assigned as the AAID if you are on an Android device, and as the IDFA if you are on an iOS device.

Reviewer: IMO, we can skip this infoblock as its already clear from the above content that for Android its AAID and for iOS its IDFA.
WDYT?

Author: Makes sense!

print(context["traits"]);
```

> For web apps, the `getRudderContext`API returns the user traits and anonymous ID only.

Reviewer: ```suggestion
For web apps, the `getRudderContext` API returns the user traits and anonymous ID only.
```
Also, acc to my understanding, the `getRudderContext`API:
* Returns **user traits** after making an identify call
* Returns **user traits** and **anonymous ID** for web apps

then why do we say:
>For web apps, the getRudderContextAPI returns the user traits and anonymous ID **only**.

Rather, it can be 
>For web apps, the getRudderContextAPI returns the user traits and anonymous ID **as well.**

LMK if I am missing anything

Author: You're right - made the change accordingly.
